Adding JavaScript to Dart page (#4000)
dart-lang/site-www: src/_guides/language/coming-from/js-to-dart.md

  such as function parameters or list items.
  This causes the formatter to place each item
  onto its own line, which helps with readability,
  especially when you have a lot of nested code
  (as can happen in Flutter layout code).
{{site.alert.end}}

For more information on using commas to make your code
read more like HTML, check out
[Using trailing commas][] on flutter.dev.

For more information about Dart conventions and linting,
check out [Effective Dart][] and [Linter rules][].

[Customizing static analysis]: /guides/language/analysis-options
[`dart fix`]: /tools/dart-fix
[Effective Dart]: /guides/language/effective-dart 
[Linter rules]: /tools/linter-rules
[Prettier]: https://prettier.io/
[Using trailing commas]: {{site.flutter}}/development/tools/formatting#using-trailing-commas

## Built-in types

While both JavaScript and Dart have types,
only Dart is a strongly typed language.
This means that all Dart types have to be
inferrable by the analyzer, explicitly defined,
or assigned as `dynamic`, which disables static type
checking for that identifier.

Dart supports nullable and non-nullable versions
of the the following built-in types:

* Numbers (`int`, `double`)

> * Strings (String`)

Reviewer: Missing leading ` on String.

Author: done

> * Booleans (bool`)

Reviewer: Missing leading ` on bool.

{{site.alert.note}}
  JavaScript has two equality operators,
  `==` and `===`. The `==` operator performs
  the equality test after doing any necessary
  type conversions. The `===` operator doesn't
  perform type conversions. Dart doesn't have
  an equivalent to `===`.
{{site.alert.end}}

For more information on these JavaScript operators,
check out [this question on Stack Overflow][].

[Built-in types]: /guides/language/language-tour#built-in-types
[Dart Language Tour]: /guides/language
[this question on Stack Overflow]: https://stackoverflow.com/questions/359494/which-equals-operator-vs-should-be-used-in-javascript-comparisons/359509#359509

For example, the equals operator `==` and the `identical()`
method are guaranteed to return true for the
same values of these types, as shown in the following code:

<!--skip-->
Dart
```dart
var a = 1;
var b = 1;

print(a == b); // Prints true
print(identical(a, b)); // Prints true
```

The next section covers basic types.
Other types are covered later in this page,
including collection types (lists, maps)
and types that aid asynchrony (futures, streams).

### Numbers

Dart has three data types for holding numbers:

> `num`

Reviewer: Should this start with ### as it's kinda acting as a header?

Author: Not a header, but thanks for pointing out the malformed html! fixed

: The equivalent to the generic number type in JavaScript.

> `int`

Reviewer: I'm guessing there should be a blank line before this to prevent this from rendering as part of the previous line

Author: ditto

: Any numeric value without a decimal point. 

> `double`

Reviewer: And a blank line before this as well

Author: ditto

final unicode = "\u{1F60E}"; // 😎,  Unicode scalar U+1F60E
```

Note that 4-digit hexadecimal values can also be used directly
(for example, `\u2665`), however, curly braces also work.
For more information on working with unicode characters,
check out [Runes and grapheme clusters][].

[Runes and grapheme clusters]: /guides/language/language-tour#characters

#### String interpolation

JavaScript supports template literals,
which are literals delimited with backtick (`\``) characters,
allowing for multiline strings,
for string interpolation with embedded expressions,
and for special constructs called tagged templates.
Dart supports string concatenation and interpolation
with embedded expressions as part of standard string literals,
meaning that you aren't required to enclose the string in
backticks to enable that functionality.
For more information,
check out [Strings][] in the [Dart Language Tour][].

[Strings]: /guides/language/language-tour#strings

As in JavaScript template literals,
insert expressions into the string literal
using the ${<expression>} syntax.
Dart expands on this by allowing the curly braces
to be omitted when the expression is a single identifier:

Dart
```dart

> var str = 'I eat \$food'; // "I eat $food".

Reviewer: Drop the \ before the $

> var str = 'I eat \${food}'; // "I eat ${food}".

Reviewer: Ditto

in both languages, and introduces Dart concepts
that are not present in vanilla JavaScript.
As a JavaScript developer,
Dart should feel quite familiar,
as both languages share many concepts. 

Like JavaScript, Dart runs on an event loop,
so both languages execute code in a similar way.
For example, asynchronous concepts like `Futures`
(`Promises` in JS) and the `async/await` syntax are very similar.

Dart is strongly typed, unlike JavaScript.
If you are also familiar with TypeScript or tooling
that adds static typing (like Flow), then learning Dart
probably won't be much of an extra step for you.
However, if you've mostly worked with vanilla JavaScript,
it might take a bit more effort to adjust.
The good news is that since Dart is strongly typed,
many errors that might exist in JavaScript code are
caught even before compiling your Dart code.

As of Dart 2.0, null safety is enabled by default,
which JavaScript doesn't support. As a JavaScript developer,
it might take a while to learn how to write null safe code,
but the trade-off is better protection against
null reference exceptions that are detected even
before compiling Dart code. (Thereby avoiding those

> dreaded `TypeErrors` that occur when doing operations

Reviewer 2: ```suggestion
dreaded `TypeError`s that occur when doing operations
```

on a JS variable that turns out to be null.)

## Conventions and linting

JavaScript and Dart both have linting tools
to enforce standard conventions. However,
while JavaScript has `ESLint` as a standalone tool
(among other available tooling),
and many different standards and configurations,
Dart has official layout conventions and includes
a linter to make compliance effortless.
To customize the lint rules for your project,
follow the [Customizing static analysis][]
instructions on dart.dev.

{{site.alert.secondary}}
  **Pro tip:** Dart provides [`dart fix`][],
  which finds and fixes errors found by the analyzer.
{{site.alert.end}}

Dart also provides a code formatter,
which is similar to JS tools like [Prettier][].
Automatically format any Dart project by running

> `dart format` on the command line. For Flutter,
> `flutter format` acts as an alias for this command.

Reviewer 2: Maybe we can link to the `dart format` documentation here like we do for `dart fix` above:

```suggestion
[`dart format`](/tools/dart-format) on the command line. 
For Flutter, `flutter format` acts as an alias for this command.
```

```

### Maps

The `Map` type in Dart is similar to the `Map` type
in JavaScript. Both types associate keys with values.
A key can be any object type, so long as all keys have
the same type; the same is true for values.
Each key occurs once at most, but you can use the
same value multiple times. 

In Dart, the dictionary is based on a hash table,
which means that keys need to be hashable. In Dart,
every object contains a unique hash.

{{site.alert.note}}
  In Dart, the hash value defaults to an instance
  of an object but you can override it to resemble
  a data class. For more information,
  check out the [`hashCode`][] property page.
{{site.alert.end}}

Here are a couple of simple `Map` examples,
created using literals:

_Dart_
```dart
final gifts = {
 'first': 'partridge',
 'second': 'turtle doves',
 'fifth': 'golden rings'
};

final nobleGases = {
 2: 'helium',
 10: 'neon',
 18: 'argon',
};
```

The following code samples provide an overview of the
basic actions that you can perform on a Dart `Map`.
The first example shows how to retrieve a value from
a `Map` using the key operator:

_Dart_
```dart
final gifts = {'first': 'partridge'};
final gift = gifts['first'];

> ```

Reviewer 2: I'm not exactly sure how to mention it here, but it seems like it may be helpful to mention how the index (`[]`) operator returns a nullable value.

`class List<E>`. Generic type `E` refers to the type of
the list’s contents. Normally, this type is automatically inferred,
which is used in some `List` class’s member types.
(For example, its first getter returns a value of type `E`.)
When defining a `List` literal,
you can explicitly define the generic type as follows:

_Dart_
```dart
// Automatic type inference
var objList = [5, 2.0]; // Type: List<num>
// Explicit type definition:
var objList = <Object>[5, 2.0]; // Type: List<Object>
// Sets work identically:
var objSet = <Object>{5, 2.0};
```

This is also true for `Map`s,
which also define their key and value types
using generics (`class Map<K, V>`):

_Dart_
```dart
// Automatic type inference
var map = {
  'foo': 'bar'
}; // Type: Map<String, String>
// Explicit type definition:
var map = <String, Object>{
  'foo': 'bar'
}; // Type: Map<String, Object>
```

## Doc comments

Regular comments work the same in Dart as they do
in JavaScript. Using `//` comments out everything beyond
it for the remaining line, and you can use `/* ... */`
to block comments spanning multiple lines.

In addition to regular comments,
Dart also has [doc comments][] that work in tandem

> with [dartdoc][]: a first party tool that generates

Reviewer 2: We've been pretty consistent in referring to the tool as `dart doc` across the website now. Which matches mentions of other tools on the page as well:

```suggestion
with [`dart doc`][]: a first party tool that generates
```

HTML documentation for Dart packages.
It's considered best practice to place doc comments
above all declarations for public members.

Define a doc comment by using three forward slashes
instead of two (`///`):

_Dart_
```dart
/// The number of characters in this chunk when unsplit.
int get length => ...
```

> [dartdoc]: /tools/dart-doc

Reviewer 2: ```suggestion
[`dart doc`]: /tools/dart-doc
```

[doc comments]: /guides/language/effective-dart/documentation#doc-comments

## Next steps

This guide has introduced you to the major differences
between Dart and JavaScript. At this point,
you might consider moving to the general documentation

> for Dart or Flutter (an open-source framework that
> uses Dart for building beautiful, natively compiled,
> multi-platform applications from a single codebase),

Reviewer 2: Maybe we can link to Flutter here:

```suggestion
for Dart or [Flutter]({{site.flutter}}) 
(an open-source framework that
uses Dart for building beautiful, natively compiled,
multi-platform applications from a single codebase),
```

where you'll find in-depth information about the
language and practical ways of getting started. 

> 

Reviewer 2: Maybe we can link to some good resources or next steps here?

Some possibilities:

- [Language tour](https://dart.dev/guides/language/language-tour) to learn more about the language
- [Library tour](https://dart.dev/guides/libraries/library-tour) to learn more about Dart’s core libraries
- [Dart codelabs](https://dart.dev/codelabs) for hands on experience learning Dart or various topics
- [Effective Dart](https://dart.dev/guides/language/effective-dart) to learn about common conventions and guidelines when writing Dart code

listening to the stream:

_Dart_
```dart
StreamSubscription subscription = stream.listen(...);
subscription.cancel();
```

This is not the only way to listen to a stream.
Similar to the `async`/`await` syntax for `Future`s,
you can combine a stream with a `for-in` loop in an
`async` context. The `for` loop invokes the
callback method for each item emitted,
and it ends when the stream completes or errors out:

_Dart_
```dart
Future<int> sumStream(Stream<int> stream) async {
  var sum = 0;
  await for (final value in stream) {
    sum += value;
  }
  return sum;
}
```

When an error occurs when listening to a stream
in this way, the error is rethrown at the line
containing the `await` keyword,
which you can handle with a `try-catch` statement:

_Dart_
```dart
try {
  await for (final value in stream) { ... }
} catch (err) {
  print('Stream encountered an error! $err');
}
```

#### Creating streams

As with `Future`s,
you have several different ways to create a stream.
The `Stream` class has utility constructors for
creating streams from `Future`s or `Iterable`s,
or for creating streams that emit values at a timed interval.
For more information, check out the [`Stream`][] API page.

> [`Stream`]: {{site.dart_pi}}/dart-async/Stream-class.html

Reviewer 2: ```suggestion
[`Stream`]: {{site.dart_api}}/dart-async/Stream-class.html
```
